Enable json output with '-json'

mjambon opened the pull request:
The `-json` option is now honored when the pattern is specified with `-e` or `-f`, that is not with a rules file. However, I didn't get rid of much code like I wanted to. Some legacy code is still used for:

* printing results as text (which is not supported for rules files)
* supporting lang=fuzzy which is not one of the semgrep languages (`Lang.t` etc.)

Anyway, I'm happy that `-json` works so we can see all the details that the text output doesn't show.

I also added pretty-printing for the json output, for user comfort. Normal semgrep usage involves large inputs and a relatively small number of matches, so the cost associated with pretty-printing should not be an issue.

Sample output:
```
~/semgrep/semgrep-core $ semgrep-core -lang java -e 'class $X' tests/java/deep_cond.java -json
{
  "matches": [
    {
      "check_id": "-e/-f",
      "path": "tests/java/deep_cond.java",
      "start": { "line": 1, "col": 1, "offset": 0 },
      "end": { "line": 1, "col": 10, "offset": 9 },
      "extra": {
        "message": "",
        "metavars": {
          "$X": {
            "start": { "line": 1, "col": 7, "offset": 6 },
            "end": { "line": 1, "col": 10, "offset": 9 },
            "abstract_content": "Foo",
            "unique_id": {
              "type": "AST",
              "md5sum": "1f0c9b0d0f3f351a3dd7a4e7ca77c675"
            }
          }
        }
      }
    }
  ],
  "errors": [],
  "stats": { "okfiles": 1, "errorfiles": 0 }
}
```

Fixes https://github.com/returntocorp/semgrep/issues/1469

Commit:
Enable json output when pattern is specified on the command line with
'-e' or 'f', and the '-json' flag is set.
Json output is now pretty-printed.
Fixes https://github.com/returntocorp/semgrep/issues/1469

mjambon commented:
Some tests fail because they're called with rules_files and without the `-json` option. Maybe `-json` should be implied by the presence of a rules file. I'll look into this after the weekend.

aryx commented:
Yes I think -rules_file implies -json.
Maybe worth adding explicitely those -json in the python wrapper

aryx commented on semgrep-core/cli/Main.ml:
```diff
@@ -626,52 +627,64 @@ let print_matches_and_errors files matches errs =
-(* simpler code path compared to semgrep_with_rules *)
-let semgrep_with_one_pattern xs =
+
+(*s: function [[Main_semgrep_core.legacy_semgrep_with_one_pattern]] *)
```
please do not modify those comments starting with s:, e:, x:, They are used by my
literate programming tool syncweb, and the tool gonna get confused by those renames.

aryx reviewed (approved):
Looks good. 
Please fix the test failures before merging though.

Commit:
Imply '-json' when passing a rules file to semgrep-core for backward
compatibility. Also, we don't support text output in this mode anyway.

Commit:
Call semgrep-core with '-json' from python.

Commit:
Don't pretty-print json output because it kills performance
significantly (not sure why, possibly due to some very
long json strings in the benchmarks)

mjambon commented:
ok, so pretty-printing the json results with `Yojson.Safe.prettify` worsened performance by an extra 50%. It's surprising. It uses ocaml's `Format` module which can be slow on some large or unusual output. I don't know if we're producing really large code snippets or something. Anyway, I turned off pretty-printing unconditionally to avoid surprises. The user should use `jq` to pretty-print json on the command line (or `ydump` which comes with the yojson library).
